sw/device/silicon_creator/lib/rescue/dfu.c

```diff
@@ -0,0 +1,274 @@
+// Copyright lowRISC contributors (OpenTitan project).
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/silicon_creator/lib/rescue/dfu.h"
+
+#include <stdbool.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/macros.h"
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/drivers/usb.h"
+#include "sw/device/silicon_creator/lib/error.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+typedef struct rescue_mode_properties {
+  uint32_t mode;
+  bool dnload;
+  bool upload;
+} rescue_mode_properties_t;
+
+static const rescue_mode_properties_t mode_by_altsetting[] = {
+    {kRescueModeFirmware, true, false},
+    {kRescueModeFirmwareSlotB, true, false},
+    {kRescueModeOpenTitanID, false, true},
+    {kRescueModeBootLog, false, true},
+    {kRescueModeBootSvcRsp, true, true},
+    {kRescueModeOwnerPage0, true, true},
+    //{ kRescueModeOwnerPage1, false, true },
+};
+
+static rom_error_t dfu_control(dfu_ctx_t *ctx, usb_setup_data_t *setup) {
+  rom_error_t result = kErrorOk;
+  if (setup->request > kDfuReqAbort) {
+    ctx->dfu_state = kDfuStateError;
+    ctx->dfu_error = kDfuErrUnknown;
+    return kErrorUsbBadSetup;
+  }
+
+  dfu_state_transition_t *tr = &dfu_state_table[setup->request][ctx->dfu_state];
+
+  // Carry out the action for this state transition.
+  switch (tr->action) {
+    case kDfuActionNone:
+      // No action: move to the next state and ack the transaction.
+      ctx->dfu_state = tr->next[0];
+      dfu_transport_data(ctx, kUsbDirIn, NULL, 0, 0);
+      break;
+    case kDfuActionStall:
+      // Stall: move to the next state and stall the transport.
+      ctx->dfu_state = tr->next[0];
+      result = kErrorUsbBadSetup;
```

Comment:
Does the DFU error value get updated somewhere for unsupported requests? I didn't spot it.

```suggestion
      ctx->dfu_state = tr->next[0];
      ctx->dfu_error = kDfuErrStalledPkt;
      result = kErrorUsbBadSetup;
```

Reply:
Does that catch requests that aren't supported _for this state_, though? (i.e. the requests are defined in the DFU spec, but they elicit the STALL response in the action / state table)

Reply:
I see what you mean.  Fixed.